### PR TITLE
perf(www): load signed renderers on demand

### PR DESCRIPTION
**Pull request by nabilfatih**
## Summary

PR #330 is the bottom layer of native stack #341. Exact head: `6e46d5fad9a965245e78a3e4f3673f577aa01fd1`.

Signed Aksara artifacts now select renderer implementations only after verification. Semantic renderers remain available, while optional custom implementations load from literal base and one authenticated renderer-domain map. Public routes, signed contracts, static rendering, Cache Components, Partial Prefetching, accessibility, styling, and responsive behavior remain product invariants.

## Why

Protected `main` measured 1,951,757 encoded JavaScript bytes on `/en` and 1,998,873 on `/id/quran/2`. The stable signed product graph after renderer and App Shell work measured 942,256 encoded bytes for Quran, about 53 percent below protected `main`. Blocking ceilings remain 1,168,654 encoded and 3,809,519 decoded for homepage, and 1,000,000 encoded and 3,100,000 decoded for Quran.

Installed Next 16.3.2 performs exact-target prefetching inside the `@next/playwright instant()` lock. Navigation therefore asserts the cached route shell while dynamic data is locked, then asserts full target content after release. The resource gate separately uses one successful router prefetch as hydration readiness. Speculative RSC failures do not poison JavaScript measurement, while every JavaScript network, HTTP, or missing-response failure remains fatal. Sources: [Next navigation](https://github.com/vercel/next.js/blob/v16.3.2/packages/next/src/client/components/segment-cache/navigation.ts), [scheduler](https://github.com/vercel/next.js/blob/v16.3.2/packages/next/src/client/components/segment-cache/scheduler.ts), and [@next/playwright](https://github.com/vercel/next.js/blob/v16.3.2/packages/next-playwright/README.md).

## Scope

- Resolve authenticated renderer requirements through typed Effect programs and literal implementation maps.
- Keep Client Components behind client-owned `next/dynamic` boundaries with SSR enabled.
- Remove caller component-map overrides and obsolete eager registries while retaining unsigned local `mdxComponents`.
- Preserve consent, Drawer, sidebar, motion, Safari list, markdown streaming, Convex provider, styling, and responsive contracts.
- Retain Playwright diagnostics only on failure for GitHub's minimum one-day period.
- No Vercel Preview, manual deployment, Gemini routing change, public API change, or Convex schema change.

## Verification at exact head

- Agent-Friendly Docs run `32820884168`: passed in 30m51s.
- Production build, signed import, renderer graph, bundle analysis, AFDocs 23/23, final runtime generation, and cleanup: passed.
- Production browser: 43/43 passed first execution in 5.7m, covering six consent scenarios, five homepage feature scenarios, 30 instant navigations, five viewports, and both JavaScript budgets. Failure upload correctly skipped.
- React Doctor run `32820884170`: passed.
- Local: www typecheck, 43-test discovery, Ultracite, boundaries, Effect source parity, architecture scans, and exact empty Gemini diff passed.

## Dependency or rollout state

- Stack order remains `#330 -> #331 -> #333`; upper layers remain blocked by this open bottom layer.
- Merge still requires one fresh exact-head Codex review, independent finding verification, and resolved threads.
- Production deployment remains protected-`main` Git integration only.

## Risks

- Numeric ceilings are Nakafa regression budgets, not framework standards.
- Failure artifacts are limited to browser output and expire after one day.

### CODE REVIEW

File: apps/www/e2e/navigation.browser.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  for (const navigationCase of navigationCases) {
+    test(`${navigationCase.name} is instant at ${viewport.name}`, async ({
+      baseURL,
+      browser,
+    }) => {
+      expect(baseURL).toBeTruthy();
+      const context = await browser.newContext({
+        hasTouch: "hasTouch" in viewport ? viewport.hasTouch : false,
+        serviceWorkers: "block",
+        viewport: { height: viewport.height, width: viewport.width },
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/navigation.browser.ts]: Fixed in commit 5e0d7f89d9. The manually created context now receives the configured Playwright baseURL, and the same validated value is passed to the hard and client navigation helper.

File: .github/workflows/agent-docs.yml [hunk 2]
```diff
@@ -217,11 +217,30 @@ jobs:
...
+          name: www-production-bundle-analysis
+          path: apps/www/.next/diagnostics/analyze
+          retention-days: 7
+
       - name: Build production app
         id: production_build
         if: env.TRUSTED_CONTENT_ENVIRONMENT == 'true'
         run: pnpm build
         env:
+          NEXT_EXPOSE_TESTING_API: "true"
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply .github/workflows/agent-docs.yml:243]: Fixed in commit 5e0d7f89d9. apps/www/turbo.json now hashes NEXT_EXPOSE_TESTING_API for the www build. Turbo dry runs report different hashes when the variable is absent versus true.

File: apps/www/e2e/support/navigation.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        marker: { kind: "heading", text: QURAN_HEADING_PATTERN },
+        name: "Quran",
+        sourceHref: "/id/quran",
+      }),
+  },
+  {
+    name: "curriculum",
+    resolve: () =>
+      Promise.resolve({
+        href: "/en/curriculum/merdeka/class-10/mathematics",
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/navigation.ts]: Fixed in commit 5e0d7f89d9. The curriculum case now discovers a visible signed-catalog subject link from the rendered homepage and verifies that discovered URL, so publication ordering can change safely.

File: apps/www/e2e/support/navigation.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  await expect(page).toHaveURL((url) => url.pathname === target.href);
+
+  if (target.marker.kind === "title") {
+    await expect(page).toHaveTitle(target.marker.text);
+    return;
+  }
+
+  const heading = page.locator("h1");
+  await expect(heading).toHaveCount(1);
+  await expect(heading).toBeVisible();
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/navigation.ts]: Fixed in commit 7bb688e0bd and preserved at exact head d09211408f. Destination assertions now select `h1:visible`, so hidden Cache Components preservation trees cannot inflate the heading count.

File: apps/www/e2e/support/resources.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+  while (true) {
+    const currentCount = yield* countJavascriptResources(page);
+    const observedAt = yield* Clock.currentTimeMillis;
+    if (currentCount !== previousCount) {
+      previousCount = currentCount;
+      lastChangeAt = observedAt;
+    }
+    if (observedAt - lastChangeAt >= RESOURCE_IDLE_MILLISECONDS) {
+      return;
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/resources.ts]: Fixed in commit d09211408f. Resource measurement now tracks matching same-origin chunk requests from request start through finish or failure, requires zero pending requests, and then observes a full one-second idle window before reading Resource Timing.

File: apps/www/e2e/support/navigation.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+      name: "curriculum",
+      sourceHref,
+    };
+  },
+};
+
+const articleCase: NavigationCase = {
+  name: "article",
+  resolve: async (page) => {
+    const sourceHref = "/en/articles/politics";
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/navigation.ts]: Fixed in commit d09211408f. Article navigation now starts at `/en/articles`, discovers a visible signed-catalog category, then discovers and verifies one visible article from that category.

**Comment by nabilfatih** [comment]
@codex review

Please review exact head `9284d685873bfab4308bda4eb1bc9454f5c43107`. Performance is the only intended product change. Styling, responsive behavior, accessibility, and interaction contracts must remain unchanged.

File: apps/www/components/marketing/about/features.tsx [hunk 7]
```diff
@@ -35,6 +35,7 @@ export async function Features({ locale }: { locale: Locale }) {
         </div>
         <FeaturesBento
           featuredTryout={featuredTryout}
+          locale={locale}
           subjectPaths={subjectPaths}
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/components/marketing/about/features.tsx:38]: Fixed at exact head `081a1285ca4776d0b43afbe1f403a5c83127c096`. The new `Features -> FeaturesBento -> child` locale prop chain is removed. `FeaturesNina` and `FeaturesProjectile` now resolve and validate the root locale locally through `getLocaleOrThrow(await getLocale())` before requesting translations.

File: packages/design-system/components/markdown/mdx.tsx [hunk 8]
```diff
@@ -121,7 +121,10 @@ export function MdxUnorderedList(props: ListProps) {
 /** Renders one MDX list item with consistent prose spacing. */
 export function MdxListItem(props: ListItemProps) {
   return (
-    <li className="space-y-4 text-pretty pl-1 leading-relaxed" {...props} />
+    <li
+      className="space-y-4 text-pretty pl-1 leading-relaxed [&>p:first-child]:inline"
+      {...props}
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/design-system/components/markdown/mdx.tsx:126]: Retained intentionally at exact head `081a1285ca4776d0b43afbe1f403a5c83127c096`. This exact selector is the coordinated Safari list-marker alignment fix from the parallel consent/UI verification: a first paragraph inside a list item must remain inline so Safari does not displace the marker. The rule is applied identically in trusted MDX and ReactMarkdown. Removing it would reintroduce the verified browser-specific visual defect. It is not part of the performance mechanism, and the design-system suite remains green with 324 tests and enforced 100 percent coverage.

File: apps/www/components/marketing/about/nina/section.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+      <div
+        className="[&>div]:scrollbar-hide relative min-h-0 flex-1 overflow-hidden"
+        role="log"
+      >
+        <div
+          style={{
+            height: "100%",
+            overflow: "auto",
+            scrollbarGutter: "stable both-edges",
+            width: "100%",
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/components/marketing/about/nina/section.tsx]: Fixed at exact head `081a1285ca4776d0b43afbe1f403a5c83127c096`. Nina again renders the existing `Conversation` and `ConversationContent` surfaces with their unchanged `initial="instant"`, `resize="smooth"`, role, scrollbar, and content classes. The static transcript remains server-owned by passing it as children through those client surfaces. Five production-browser scenarios now assert the initial bottom position and disclosure behavior at all target viewports.

File: apps/www/e2e/featured-renderer.mts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  const selectedRenderers = yield* selectRendererImplementations(
+    item.artifact.payload
+  );
+  const selectedCustomNames = selectedRenderers
+    .filter(({ kind }) => kind === "implementation")
+    .map(({ name }) => name)
+    .sort();
+  const signedCustomNames = item.artifact.payload.requiredComponents
+    .map(({ name }) => name)
+    .sort();
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/featured-renderer.mts]: Fixed at exact head `081a1285ca4776d0b43afbe1f403a5c83127c096`. The assertion now compares every selected renderer name with every authenticated `requiredComponents` name, so signed semantic names and physical implementations are treated symmetrically. The exact-head www suite passes 1,355 tests with enforced 100 percent coverage.

File: apps/www/components/marketing/about/projectile/client.tsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+/**
+ * Imports the scene module after viewport intent.
+ *
+ * @see https://nextjs.org/docs/app/guides/lazy-loading
+ */
+const ProjectileScene = dynamic(() =>
+  import("@/components/marketing/about/projectile/scene").then(
+    ({ ProjectileScene: Scene }) => Scene
+  )
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/components/marketing/about/projectile/client.tsx:47]: Fixed at exact head `081a1285ca4776d0b43afbe1f403a5c83127c096`. The client-owned `next/dynamic` boundary now runs the restored `loadProjectileScene` Effect around the literal scene import. It preserves the original four total attempts with exponential delay and reports terminal failures before React and Next.js receive the rejection. The restored loader and its two retry/failure tests are byte-identical to current main. Five production-browser scenarios also verify intersection deferral, canvas activation, and scenario interaction across every target viewport.

**Comment by nabilfatih** [comment]
@codex review

Please review exact head `081a1285ca4776d0b43afbe1f403a5c83127c096`. The five prior threads are resolved with code and test evidence. Treat styling, responsive behavior, accessibility, Drawer/Dialog behavior, Nina scroll semantics, projectile retries, signed renderer ownership, Cache Components, Partial Prefetching, and resource budgets as blocking contracts.

**Comment by nabilfatih** [comment]
@codex review

Please review exact head `50b40254d6efb472b76085dd4d0d4391efe05910`. Treat styling, responsive behavior, accessibility, signed renderer integrity, Cache Components, Partial Prefetching, and the no-Vercel-Preview policy as blocking contracts.

File: apps/www/components/marketing/about/projectile/client.tsx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        ({ ProjectileScene: Scene }) => Scene
+      )
+    ).pipe(
+      Effect.tapError((error) =>
+        reportClientException(error, {
+          operation: "load-projectile-scene",
+          source: "home-features",
+        })
+      )
+    )
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/components/marketing/about/projectile/client.tsx:46]: Fixed at exact head `c274cc396550a4eb1bd951c9119ba1b38d0a0fc8`. The four-attempt Effect loader now uses `Effect.matchEffect`: terminal failure is reported, then resolved to an empty component for this optional WebGL surface, so the route boundary is not rejected.

File: apps/www/e2e/features.browser.ts [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+)(function* (page: Page) {
+  const visual = page.getByRole("region", {
+    name: "Cannonball projectile analysis visual",
+  });
+  const canvas = visual.locator("canvas");
+  const highArc = page.getByRole("button", { name: "High Arc" });
+
+  yield* Effect.promise(() => expect(visual).toHaveCount(1));
+  yield* Effect.promise(() => expect(canvas).toHaveCount(0));
+  yield* Effect.promise(() => visual.scrollIntoViewIfNeeded());
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/features.browser.ts:142]: Fixed at exact head `c274cc396550a4eb1bd951c9119ba1b38d0a0fc8`. A dedicated deferral assertion now runs immediately after initial navigation and before layout measurement, Nina scrolling, or projectile scrolling. The later helper only verifies activation and interaction.

File: apps/www/components/providers/convex.tsx [hunk 14]
```diff
@@ -119,7 +112,22 @@ function useBetterAuth() {
...
+  children: ReactNode;
+  convexUrl: string;
+}) {
+  const convex = useMemo(
+    () =>
+      new ConvexReactClient(convexUrl, {
+        initialAuthTokenReuse: true,
+        verbose: false,
+      }),
+    [convexUrl]
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/components/providers/convex.tsx]: Fixed at exact head `c274cc396550a4eb1bd951c9119ba1b38d0a0fc8`. The provider now retains one deliberate page-lifetime `ConvexReactClient` singleton instead of constructing provider-owned clients during render. This follows the installed Convex React usage contract and prevents discarded renders or auth/app layout transitions from abandoning clients.

File: apps/www/components/analytics/consent/controls.tsx [hunk 15]
```diff
@@ -53,17 +63,20 @@ export function AnalyticsConsentControls() {
...
-        setOpen={setPreferencesOpen}
-        title={t("title")}
-      >
-        <div className="flex flex-col gap-3">
-          <ConsentDetails />
-        </div>
-      </ResponsiveDialog>
+      {preferences.isOpen ? (
+        <ConsentPreferences
+          description={t(`status-${preferences.statusAtOpen}`)}
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/components/analytics/consent/controls.tsx]: Fixed at exact head `c274cc396550a4eb1bd951c9119ba1b38d0a0fc8`. The deferred consent experiment is removed completely. `ResponsiveDialog` is synchronously mounted again, the shared primitive and all consent trigger behavior are byte-identical to protected `main`, and the temporary preload module is deleted.

**Comment by nabilfatih** [comment]
@codex review exact head `c274cc396550a4eb1bd951c9119ba1b38d0a0fc8`. Please re-audit the full diff for correctness, UX and styling preservation, Effect architecture, renderer loading contracts, and production-browser gate reliability. Do not treat earlier heads as current.

**Comment by nabilfatih** [comment]
@codex review exact rebased head `6f520e49ee137394db32d5a31d0ccb0300535012` on protected base `fcdc35f1fcc11d67780b69512360a41dc97392ca`. Please audit the complete current diff, including UX and styling preservation, Effect architecture, renderer loading contracts, and production-browser gate reliability. Earlier-head reviews are not current.

**Comment by nabilfatih** [comment]
@codex review exact head `22fa850d5bc4cbcb04bfc0ee798c81ea72461556` on protected base `fcdc35f1fcc11d67780b69512360a41dc97392ca`. Please audit the complete diff, with particular attention to the Effect-native Playwright navigation seam, its typed missing-link failure, the single test-boundary runner, installed `@next/playwright` callback contract, UX and styling preservation, renderer loading integrity, and production-browser gate reliability. Earlier-head reviews are not current.

**Comment by nabilfatih** [comment]
@codex review exact head `f4c58595f1b4b872905dffd46e528379d514ce61` on protected base `fcdc35f1fcc11d67780b69512360a41dc97392ca`. Please audit the complete current diff, especially the source-backed signed normal-prefetch budget contract, Base UI focus containment polling, wide-layout reduced-motion browser coverage, Effect-native E2E boundaries, renderer integrity, and preservation of styling, responsive behavior, Drawer and Dialog UX, consent semantics, Cache Components, Partial Prefetching, and the no-Vercel-Preview policy. Earlier-head reviews are not current.

File: apps/www/e2e/support/resources.ts [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+      };
+      const handleRequestSettled = (request: Request) => {
+        if (!pendingRequests.delete(request)) {
+          return;
+        }
+        revision += 1;
+      };
+
+      page.on("request", handleRequest);
+      page.on("requestfailed", handleRequestSettled);
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/resources.ts]: Fixed in exact commit `75b3c0af76a0bb8fe6aec6e41ea3ea6a595ee1de`. Matching JavaScript and router-prefetch requests now remain pending through settlement. `requestfailed` records the typed network outcome with optional `request.failure()?.errorText`; `requestfinished` records non-OK HTTP and missing-response outcomes. The Effect fails before returning any timing run, including a second failure and revision check after the Resource Timing snapshot.

File: apps/www/e2e/support/resources.ts [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+    ) {
+      previousCount = currentCount;
+      previousRevision = requestTracker.revision;
+      lastChangeAt = observedAt;
+    }
+    if (
+      requestTracker.pendingCount === 0 &&
+      observedAt - lastChangeAt >= RESOURCE_IDLE_MILLISECONDS
+    ) {
+      return;
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/resources.ts]: Fixed in exact commit `75b3c0af76a0bb8fe6aec6e41ea3ea6a595ee1de`. The idle window cannot begin until the tracker observes a same-origin request carrying installed Next.js 16.3.2's `next-router-prefetch` header. Router-prefetch and JavaScript requests are both tracked as pending, and the timing snapshot is returned only if no request or revision changed while it was read.

**Comment by nabilfatih** [comment]
@codex review exact head `75b3c0af76a0bb8fe6aec6e41ea3ea6a595ee1de`. Please review the complete PR at this exact commit, including the fail-closed Resource Timing tracker, Next.js prefetch readiness contract, Effect-native E2E boundaries, UX and styling preservation, and the updated evidence in the PR body.

**Comment by nabilfatih** [comment]
@codex review

File: apps/www/playwright.config.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+export default defineConfig({
+  expect: {
+    timeout: 5000,
+  },
+  forbidOnly: Boolean(process.env.CI),
+  fullyParallel: false,
+  outputDir: "../../.cache/playwright/www",
+  reporter: "list",
+  retries: process.env.CI ? 1 : 0,
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/playwright.config.ts:14]: Fixed at exact commit `d7d34639a8f32e77b2108c15550b670a1349750e`. CI now sets installed Playwright 1.62.1's documented `failOnFlakyTests` option. The existing single retry remains only for diagnostic artifacts, while Playwright's runner returns failure for any flaky classification. WWW typecheck, lint across 2,910 files, and CI-mode discovery of all 43 browser checks passed.

**Comment by nabilfatih** [comment]
@codex review

**Comment by nabilfatih** [comment]
@codex review

File: apps/www/e2e/support/navigation.ts [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+
+const navigateClient = Effect.fn("NakafaE2E.navigateClient")(function* (
+  page: Page,
+  target: NavigationTarget,
+  hasTouch: boolean
+) {
+  const sourceResponse = yield* Effect.promise(() =>
+    page.goto(target.sourceHref, { waitUntil: "domcontentloaded" })
+  );
+  yield* Effect.sync(() => expect(sourceResponse?.ok()).toBe(true));
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/navigation.ts]: Fixed at exact head `c551a535a43dbefd84887f3ede6db6352f732e73`. Client readiness now waits for committed Next App Router history state, brings the exact Link into the viewport, and observes one successful settled same-origin exact-target `next-router-prefetch` before entering installed `@next/playwright` `instant()`. The existing request tracker owns the sole Playwright request lifecycle and schema-derived failure payload; no second tracker or quiet-window completion claim remains. Local proof passed www typecheck, 203 files and 1,356 tests at 100 percent coverage, full lint, boundaries, Effect RC110 source parity, all 43 Playwright cases listed, diff integrity, and the no-raw-try/catch and single-lifecycle scans. Exact-head signed CI remains blocking.

File: apps/www/playwright.config.ts [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+  testDir: "./e2e",
+  testMatch: "**/*.browser.ts",
+  timeout: 120_000,
+  use: {
+    baseURL,
+    screenshot: "only-on-failure",
+    serviceWorkers: "block",
+    trace: "retain-on-failure",
+    video: "retain-on-failure",
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/playwright.config.ts:23]: Confirmed valid. Run 32812646537 named screenshots, traces, videos, and error-context paths under `.cache/playwright/www`, but GitHub retained only the analyzer artifact after the runner cleanup. A bounded failure-only Playwright artifact upload is required. It is deliberately separate from the navigation-semantic checkpoint, so this thread remains open until the CI change and retention proof are pushed.

File: apps/www/e2e/support/request-tracker.ts [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+          const pendingRequest = settleRequest(request);
+          if (!pendingRequest) {
+            return;
+          }
+          const requestFailure = request.failure();
+          failure ??= requestFailure
+            ? {
+                ...pendingRequest.details,
+                errorText: requestFailure.errorText,
+                outcome: "network",
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/request-tracker.ts]: Confirmed valid from exact run 32812646537. Every homepage attempt recorded the same `/en` prefetch abort, and the Quran resource run recorded the analogous `/id` abort. Exact head `bddeed45dad63d80cc22614db90b7b0fd40d7063` removes the navigation tracker and its external exact-target precondition, then asserts the target shell inside the installed `instant()` lock. WWW typecheck, all 43-case discovery, touched-file lint, Effect source parity, and architecture scans pass. The resource classification remains intentionally unchanged and this thread stays open until that route-transition case is source-classified and fixed.

[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/e2e/support/request-tracker.ts]: Verified and corrected at exact head `c59903d87179e9c0546efce27e21a31d1bb7c07d`. Installed Next 16.3.2 does not support the review's specific causal claim: `cancelPrefetchTask` marks and dequeues a task, and the scheduler explicitly says it does not abort in-progress requests. The observed `/id?_rsc=...`, header `3`, `net::ERR_ABORTED` request is nevertheless a speculative RSC App Shell prefetch, not a `/_next/static/...js` resource.

Commit `cf5a4c529560cc9e3244159e987b6d9e12da4ffb` now stores outcomes per request kind. Only `getFailure("javascript")` can fail the byte measurement. Readiness requires `successfulCount("prefetch") > 0` within the existing 15-second bound; if none succeeds, the typed timeout preserves the actual prefetch failure fields. Required JavaScript network, HTTP, and missing-response failures remain fatal before and after the timing snapshot.

Proof at the exact head: www typecheck, touched-file Ultracite, 43-case Playwright discovery, Effect RC110 source parity, clean E2E try/catch and `waitForTimeout` scans, plus three uncached live Quran measurements through this seam. Exact-head production acceptance remains blocking.

File: apps/www/playwright.config.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (nabilfatih) [reply apps/www/playwright.config.ts:23]: Fixed at exact head `c59903d87179e9c0546efce27e21a31d1bb7c07d` in commit `c59903d87179e9c0546efce27e21a31d1bb7c07d`. The workflow now gives the browser step a stable ID and uploads only `.cache/playwright/www` when that exact step fails. The artifact name includes `github.run_attempt`, uses the existing reviewed `actions/upload-artifact@v7.0.1` immutable SHA, requires files to exist, and retains them for GitHub's minimum one day. Successful browser runs upload nothing.

The canonical TypeScript Action policy now expects both reviewed uses and retains the same pin. `pnpm test:scripts` passed all 9 dependency, Action policy, usage-count, and Effect-source tests; script typecheck and diff integrity also pass. This follows GitHub's documented test-artifact use and custom retention contract while limiting storage accrual. Exact-head CI will provide the end-to-end workflow proof.

**Comment by nabilfatih** [comment]
@codex review
